File: go/scanner/module/http/directory.go
package http

import (
	"fmt"
	"io/ioutil"
	"net/http"
	"regexp"

	"github.com/bloom42/phaser/common/phaser"
	"github.com/bloom42/phaser/common/phaser/findings"
	"github.com/bloom42/phaser/scanner/module"
)

type DirectoryListingInformationDisclosure struct{}

func (DirectoryListingInformationDisclosure) Name() string {
	return "http/directory_listing_information_disclosure"
}

func (DirectoryListingInformationDisclosure) Description() string {
	return "Check for enabled direstory listing, which often leak information"
}

func (DirectoryListingInformationDisclosure) Author() string {
	return "Sylvain Kerkour <sylvain@kerkour.com>"
}

func (DirectoryListingInformationDisclosure) Version() string {
	return "0.1.0"
}

const DIRECTORY_LISTING_REGEXP = "<title>Index of .*<\\/title>"

func (DirectoryListingInformationDisclosure) Run(scan *phaser.Scan, target *phaser.Target, port phaser.Port) (module.Result, []error) {
	errs := []error{}
	var ret module.Result
	protocol := "http"

	if !port.HTTP && !port.HTTPS {
		return ret, errs
	}
	if port.HTTPS {
		protocol = "https"
	}

	URL := fmt.Sprintf("%s://%s:%d/", protocol, target.Host, port.ID)
	req, err := http.NewRequest("GET", URL, nil)
	if err != nil {
		errs = append(errs, err)
		return ret, errs
	}
	req.Header.Set("User-Agent", scan.Profile.HTTP.UserAgent)
	res, err := scan.HTTPClient.Do(req)
	if err != nil {
		errs = append(errs, err)
		return ret, errs
	}
	// response, _ := httputil.DumpResponse(res, true)
	body, err := ioutil.ReadAll(res.Body)
	if err != nil {
		errs = append(errs, err)
		return ret, errs
	}
	res.Body.Close()
	bodyStr := string(body)

	matched, err := regexp.MatchString(DIRECTORY_LISTING_REGEXP, bodyStr)
	if err != nil {
		errs = append(errs, err)
		return ret, errs
	}

	if matched {
		ret = findings.URL{URL: URL}
	}

	return ret, errs
}

File: src/scanner/modules/http/directory.rs
use crate::scanner::{
    module,
    findings,
    Scan,
    Target,
    TargetKind,
};
use std::process::{Command};
use regex::Regex;

pub struct DirectoryListing{}

impl module::BaseModule for DirectoryListing {
    fn name(&self) -> String {
        return String::from("http/directory-listing");
    }

    fn description(&self) -> String {
        return String::from("Check for enabled direstory listing, which often leak informationr");
    }

    fn author(&self) -> String {
        return String::from("Sylvain Kerkour <sylvain@kerkour.com>")
    }

    fn version(&self) -> String {
        return String::from("0.1.0");
    }
}

impl module::PortModule for DirectoryListing {
    fn run(&self, _: &Scan, target: &Target, port: &findings::Port) -> (Option<findings::Data>, Vec<String>) {
        let mut errs = vec!();
        let mut ret = None;

        let protocol = if port.http {
            "http"
        } else if port.https {
            "https"
        } else {
            ""
        };

        if protocol == "" {
            return (ret, errs);
        }

        let url = format!("{}://{}:{}/", &protocol, &target.host, &port.id);
        let body = reqwest::get(&url)
            .expect("error fetching url for direcotry listing")
            .text()
            .expect("error getting body to txt");

        let re = Regex::new(r"<title>Index of .*</title>").expect("compiling regexp");

        if re.is_match(&body) {
            ret = Some(findings::Data::Url(findings::Url{
                url,
            }));
        }

        return (ret, errs);
    }
}


#[cfg(test)]
mod tests {
    // Note this useful idiom: importing names from outer (for mod tests) scope.
    use super::*;

    #[test]
    fn test_regexp() {
        let re = Regex::new(r"<title>Index of .*</title>").expect("compiling regexp");
        let body = "lol lol lol ol ol< LO> OL  <title>Index of kerkour.com</title> sdsds".to_string();
        let body2 = "lol lol lol ol ol< LO> OL  <tle>Index of kerkour.com</title> sdsds".to_string();
        let body3 = "".to_string();
        let body4 = "lol lol lol ol ol< LO> OL  <title>Index</title> sdsds".to_string();
        assert_eq!(true, re.is_match(&body));
        assert_eq!(false, re.is_match(&body2));
        assert_eq!(false, re.is_match(&body3));
        assert_eq!(false, re.is_match(&body4));
    }
}
